### PR TITLE
#1217 - operators for option type

### PR DESCRIPTION
**Pull request by jczuchnowski**
Part of work for https://github.com/zio/zio/issues/1217 . In short - improve working with options:
```scala
    val value1: Task[Option[Int]] = Task(Some(1))
    val value2: Task[Option[Int]] = Task(Some(2))

    val result: IO[Option[Throwable], Int] = for {
      val1 <- value1.some
      val2 <- value2.some
    } yield val1 + val2
        
    val resultOption: Task[Option[Int]] = result.peelError
```

### CODE REVIEW

**Comment by adamgfraser** [comment]
In case it is helpful, here is the work I did on implementing `none` and `fuseError`.

```
  def none[B](implicit ev: A <:< Option[B]): ZIO[R, Option[E], Unit] =
    self.foldM(
      e => ZIO.fail(Some(e)),
      a => ev(a).fold[ZIO[R, Option[E], Unit]](ZIO.succeed(()))(_ => ZIO.fail(None))
    )
```

```
  def fuseError[E1, E2 <: E1](default: E2)(implicit ev: E <:< Option[E1]): ZIO[R, E1, A] =
    self.mapError(e => ev(e).getOrElse(default))
```

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -395,6 +395,12 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   final def get[E1 >: E, B](implicit ev1: E1 =:= Nothing, ev2: A <:< Option[B]): ZIO[R, Unit, B] =
     ZIO.absolve(self.mapError(ev1).map(ev2(_).toRight(())))
 
+  final def some[B](implicit ev: A <:< Option[B]): ZIO[R, Option[E], B] = 
+    self.foldM(e => ZIO.fail(Some(e)), a => a.fold[ZIO[R, Option[E], B]](ZIO.fail(Option.empty[E]))(ZIO.succeed(_).mapError(_ => Option.empty[E])))
+
+  final def peelError[E1](implicit ev: E <:< Option[E1]): ZIO[R, E1, Option[A]] =
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/ZIO.scala]: Think we should call this `optional`. Better name than `peelError`, anyway. Otherwise looks good!

**Comment by jczuchnowski** [comment]
@jdegoes @adamgfraser what are the actual use cases for `none` and `fuseError`. I'm trying to come up with a meaningful test cases.

**Comment by adamgfraser** [comment]
@jczuchnowski `none` is basically asserting that the optional value in a `ZIO` is empty and failing if it is not. So you should have:

```scala
IO.succeed(None).none === IO(())
IO.succeed(Some(1)).none === IO.fail(Option.empty)
IO.fail("Fail").none === IO.fail(Some("Fail"))
```

Fuse error is taking a `ZIO` that could fail with 'Some(error)' or 'None' and adding a default error value to fail with to replace the `None` case. So you should have:

```scala
IO.fail(Some("Fail")).fuseError("Default") === IO.fail("Fail")
IO.fail(None).fuseError("Default") === IO.fail("Default")
```

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -400,6 +400,9 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   final def forkOn(ec: ExecutionContext): ZIO[R, E, Fiber[E, A]] =
     self.on(ec).fork
 
+  final def fuseError[E1, E2 <: E1](default: E2)(implicit ev: E <:< Option[E1]): ZIO[R, E1, A] =
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/ZIO.scala]: How about we call this `flattenErrorOption`?

Comment (jczuchnowski) [reply core/shared/src/main/scala/zio/ZIO.scala]: Definitely more descriptive.

**Pull request merged**